File: leetcode/Remove Duplicates from Sorted Array II/Remove Duplicates from Sorted Array II.java
public class Solution {
    public int removeDuplicates(int[] nums) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        int dupCount = 0;
        for (int i = 0; i < nums.length; ++i) {
            Integer val = map.get(nums[i]);
            if (val == null) {
                map.put(nums[i], 1);
            } else {
                int count = val.intValue();
                if (count == 1) {
                    map.put(nums[i], 2);
                } else if (count == 2) {
                    dupCount++;
                }
            }
            if (dupCount != 0) {
                nums[i - dupCount] = nums[i];
            }
        }
        return nums.length - dupCount;
    }
}

File: leetcode/Remove Duplicates from Sorted Array II/Remove Duplicates from Sorted Array II.cpp
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        set<int> s1, s2;
        int size = nums.size();
        int dupcount = 0;
        for (int i = 0; i < size; ++i) {
            if (s1.count(nums.at(i))) {
               if (s2.count(nums.at(i))) {
                   dupcount++;
               } else {
                   s2.insert(nums.at(i));
                   nums.at(i - dupcount) = nums.at(i);
               }
            } else {
                s1.insert(nums.at(i));
                nums.at(i - dupcount) = nums.at(i);
            }
        }
        return s1.size() + s2.size();
    }
};
